src/typetree.jl

```diff
@@ -143,13 +143,27 @@ else
     ismutabletype(T) = isa(T, DataType) && T.mutable
 end
 
+const AlreadyWarned = Dict{Symbol,Bool}()
+const AlreadyWarnedLock = Threads.SpinLock()
+
+function recursive_type_warning(@nospecialize(T))
+    haskey(AlreadyWarned, Symbol(T)) && return nothing
+    lock(AlreadyWarnedLock) do
+        haskey(AlreadyWarned, Symbol(T)) && return nothing
+        GPUCompiler.@safe_warn "Recursive type" T
+        AlreadyWarned[Symbol(T)] = true
+        return nothing
+    end
+    return nothing
+end
+
 function typetree(@nospecialize(T), ctx, dl, seen=nothing)
     if T isa UnionAll || T isa Union || T == Union{} || Base.isabstracttype(T)
         return TypeTree()
     end
 
     if seen !== nothing && T ∈ seen
-        GPUCompiler.@safe_warn "Recursive type" T
+        recursive_type_warning(T)
```

Comment:
Maybe we could simply demote the warning to a debug statement? It's not like something the user will be able to do anything about it

Reply:
I'm also debating if we just get rid of the warning. The only implication of it is that we don't fill out a recursive type tree, so slightly less info.

Reply:
FWIW I haven't had issues with recursive types despite the warning.

Feel free to edit my PR directly by the way.